### PR TITLE
NGG coding rework: Rewrite primitive shader LDS management

### PR DESCRIPTION
**Pull request by amdrexu**

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
0f9076a4241b9a042c03bf74b67b835f8168c49d Jenkins build error.
/home/amdvlk/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/include/llvm-dialects/TableGen/Format.h:77:9: error: ‘Optional’ in namespace ‘llvm’ does not name a template type
   llvm::Optional<llvm::StringRef> getSubstFor(PHKind placeholder) const;
         ^~~~~~~~
/home/amdvlk/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/include/llvm-dialects/TableGen/Format.h:78:9: error: ‘Optional’ in namespace ‘llvm’ does not name a template type
   llvm::Optional<llvm::StringRef> getSubstFor(llvm::StringRef placeholder) const;
         ^~~~~~~~
/home/amdvlk/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/lib/TableGen/Format.cpp:63:1: error: ‘Optional’ does not name a type
 Optional<StringRef>
 ^~~~~~~~
/home/amdvlk/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/lib/TableGen/Format.cpp:74:1: error: ‘Optional’ does not name a type
 Optional<StringRef> FmtContext::getSubstFor(StringRef placeholder) const {
 ^~~~~~~~
/home/amdvlk/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/lib/TableGen/Format.cpp: In member function ‘void llvm_dialects::FmtObjectBase::format(llvm::raw_ostream&) const’:
/home/amdvlk/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/lib/TableGen/Format.cpp:191:9: error: ‘Optional’ was not declared in this scope
         Optional<StringRef> subst;
         ^~~~~~~~
/home/amdvlk/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/lib/TableGen/Format.cpp:191:27: error: expected primary-expression before ‘>’ token
         Optional<StringRef> subst;
                           ^
/home/amdvlk/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/lib/TableGen/Format.cpp:191:29: error: ‘subst’ was not declared in this scope
         Optional<StringRef> subst;
                             ^~~~~
/home/amdvlk/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/lib/TableGen/Format.cpp:191:29: note: suggested alternative: ‘strstr’
         Optional<StringRef> subst;
                             ^~~~~
--
/home/amdvlk/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/lib/TableGen/Format.cpp:194:28: error: ‘const class llvm_dialects::FmtContext’ has no member named ‘getSubstFor’
           subst = context->getSubstFor(repl.spec.substr(1));
                            ^~~~~~~~~~~
/home/amdvlk/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/lib/TableGen/Format.cpp:196:28: error: ‘const class llvm_dialects::FmtContext’ has no member named ‘getSubstFor’
           subst = context->getSubstFor(repl.placeholder);
                            ^~~~~~~~~~~
/home/amdvlk/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/lib/TableGen/Format.cpp:211:47: warning: ‘llvm::ArrayRef<T> llvm::makeArrayRef(const std::vector<T>&) [with T = llvm::detail::format_adapter*]’ is deprecated: Use deduction guide instead [-Wdeprecated-declarations]
       auto range = llvm::makeArrayRef(adapters);
                                               ^
--
/home/amdvlk/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/include/llvm-dialects/TableGen/Format.h:77:9: error: ‘Optional’ in namespace ‘llvm’ does not name a template type
   llvm::Optional<llvm::StringRef> getSubstFor(PHKind placeholder) const;
         ^~~~~~~~
/home/amdvlk/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/include/llvm-dialects/TableGen/Format.h:78:9: error: ‘Optional’ in namespace ‘llvm’ does not name a template type
   llvm::Optional<llvm::StringRef> getSubstFor(llvm::StringRef placeholder) const;
         ^~~~~~~~
/home/amdvlk/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/lib/TableGen/GenDialect.cpp: In member function ‘std::string {anonymous}::SymbolTable::chooseName(llvm::StringRef)’:
/home/amdvlk/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/lib/TableGen/GenDialect.cpp:575:40: warning: ‘llvm::ArrayRef<U> llvm::makeArrayRef(const T&) [with T = llvm::StringRef]’ is deprecated: Use deduction guide instead [-Wdeprecated-declarations]
     return chooseName(makeArrayRef(name));

**Comment by amdvlk-admin** [comment]
### Test summary for commit abfd15a3b10e2418301031fc244d9e682ef4cad6
<details><summary>CTS tests (Failed: 37/282025)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 38825/70507 (55.1%)</li>
<li><details><summary>Failed: 37/70507 (0.1%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.transform_feedback.simple.host_query_reset_line_list_0_250_64bits
Stack trace: Fail

FAILURE: dEQP-VK.transform_feedback.simple.host_query_reset_line_list_with_adjacency_0_248_32bits
Stack trace: Fail

FAILURE: dEQP-VK.transform_feedback.simple.host_query_reset_line_strip_0_127_64bits
Stack trace: Fail

FAILURE: dEQP-VK.transform_feedback.simple.host_query_reset_line_strip_0_251_32bits
Stack trace: Fail

FAILURE: dEQP-VK.transform_feedback.simple.host_query_reset_line_strip_0_509_64bits
Stack trace: Fail

FAILURE: dEQP-VK.transform_feedback.simple.host_query_reset_line_strip_with_adjacency_0_251_64bits
Stack trace: Fail

FAILURE: dEQP-VK.transform_feedback.simple.host_query_reset_triangle_fan_0_251_64bits
Stack trace: Fail

FAILURE: dEQP-VK.transform_feedback.simple.host_query_reset_triangle_fan_0_61_32bits
Stack trace: Fail

FAILURE: dEQP-VK.transform_feedback.simple.host_query_reset_triangle_list_0_249_64bits
Stack trace: Fail

FAILURE: dEQP-VK.transform_feedback.simple.host_query_reset_triangle_strip_0_509_64bits
Stack trace: Fail

FAILURE: dEQP-VK.transform_feedback.simple.query_0_251_32bits_endqueryindexed_streamid_0
Stack trace: Fail

FAILURE: dEQP-VK.transform_feedback.simple.query_0_251_64bits
Stack trace: Fail

FAILURE: dEQP-VK.transform_feedback.simple.query_copy_line_list_0_508_64bits
Stack trace: Fail

FAILURE: dEQP-VK.transform_feedback.simple.query_copy_line_list_with_adjacency_0_248_64bits
Stack trace: Fail

FAILURE: dEQP-VK.transform_feedback.simple.query_copy_line_list_with_adjacency_0_508_64bits
Stack trace: Fail

FAILURE: dEQP-VK.transform_feedback.simple.query_copy_line_strip_0_251_32bits
Stack trace: Fail

FAILURE: dEQP-VK.transform_feedback.simple.query_copy_triangle_fan_0_251_64bits
Stack trace: Fail

FAILURE: dEQP-VK.transform_feedback.simple.query_copy_triangle_fan_0_61_32bits
Stack trace: Fail

FAILURE: dEQP-VK.transform_feedback.simple.query_copy_triangle_strip_0_127_32bits
Stack trace: Fail

FAILURE: dEQP-VK.transform_feedback.simple.query_line_list_0_250_64bits_endqueryindexed_streamid_0
Stack trace: Fail
...
```
</p></details></li>

<li>Not Supported: 31645/70507 (44.9%)</li><li>Warnings: 0/70507 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi2x</summary><ul>

<li>Passed: 39112/70507 (55.5%)</li>
<li>Failed: 0/70507 (0.0%)</li>
<li>Not Supported: 31395/70507 (44.5%)</li><li>Warnings: 0/70507 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35959/70509 (51.0%)</li>
<li>Failed: 0/70509 (0.0%)</li>
<li>Not Supported: 34550/70509 (49.0%)</li><li>Warnings: 0/70509 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 35861/70502 (50.9%)</li>
<li>Failed: 0/70502 (0.0%)</li>
<li>Not Supported: 34641/70502 (49.1%)</li><li>Warnings: 0/70502 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 08a48634750109515834fbb4d1fffa3cb2c4c604
<details><summary>CTS tests (Failed: 0/282025)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 38862/70507 (55.1%)</li>
<li>Failed: 0/70507 (0.0%)</li>
<li>Not Supported: 31645/70507 (44.9%)</li><li>Warnings: 0/70507 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi2x</summary><ul>

<li>Passed: 39112/70507 (55.5%)</li>
<li>Failed: 0/70507 (0.0%)</li>
<li>Not Supported: 31395/70507 (44.5%)</li><li>Warnings: 0/70507 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35959/70509 (51.0%)</li>
<li>Failed: 0/70509 (0.0%)</li>
<li>Not Supported: 34550/70509 (49.0%)</li><li>Warnings: 0/70509 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 35861/70502 (50.9%)</li>
<li>Failed: 0/70502 (0.0%)</li>
<li>Not Supported: 34641/70502 (49.1%)</li><li>Warnings: 0/70502 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — COMMENTED** [review]
Looks reasonable, just some small comments inline.

File: lgc/patch/NggPrimShader.h [hunk 1]
```diff
@@ -160,14 +158,42 @@ struct XfbOutputExport {
...
+// Map: LDS region -> <region Offset, region Size>
+typedef std::unordered_map<PrimShaderLdsRegion, std::pair<unsigned, unsigned>> PrimShaderLdsLayout;
+
 // =====================================================================================================================
 // Represents the manager of NGG primitive shader.
 class NggPrimShader {
 public:
   NggPrimShader(PipelineState *pipelineState);
-  ~NggPrimShader();
+  ~NggPrimShader(){};
```
[the remark this answers is not in the document]
Comment (amdrexu) [reply lgc/patch/NggPrimShader.h:192]: Removed. Thank you.

File: lgc/patch/NggPrimShader.cpp [hunk 2]
```diff
@@ -151,18 +149,336 @@ unsigned NggPrimShader::calcEsGsRingItemSize(PipelineState *pipelineState) {
...
+
+  // XFB statistics
+  ldsRegionSize =
+      MaxTransformFeedbackBuffers + 1; // 1 dword per XFB buffer: dword written, 1 dword: primitives to write
+  if (ldsLayout) {
+    printLdsRegionInfo("XFB Statistics", ldsOffset, ldsRegionSize);
+    (*ldsLayout)[PrimShaderLdsRegion::XfbStats] = std::make_pair(ldsOffset, ldsRegionSize);
+    ldsOffset += ldsRegionSize;
+  }
+  ldsUsageInfo.esExtraLdsSize += ldsRegionSize;
```
[the remark this answers is not in the document]
Comment (amdrexu) [reply lgc/patch/NggPrimShader.cpp:448]: Yes, I will add the conditional check.

File: lgc/patch/NggPrimShader.cpp [hunk 3]
```diff
@@ -151,18 +149,336 @@ unsigned NggPrimShader::calcEsGsRingItemSize(PipelineState *pipelineState) {
...
+      }
+      ldsUsageInfo.gsExtraLdsSize += ldsRegionSize;
+    }
+
+    // Vertex index map (compacted -> uncompacted)
+    if (pipelineState->getNggControl()->compactVertex) {
+      if (pipelineState->enableSwXfb()) {
+        if (ldsLayout) {
+          // NOTE: If SW emulated stream-out is enabled, this region is overlapped with PrimitiveIndexMap
+          (*ldsLayout)[PrimShaderLdsRegion::VertexIndexMap] = (*ldsLayout)[PrimShaderLdsRegion::VertexIndexMap];
```
[the remark this answers is not in the document]
Comment (amdrexu) [reply lgc/patch/NggPrimShader.cpp:271]: Oh. It is a typo. Good catch. I didn't check vertex compaction with XFB on Navi3x. So this issue is neglected.

**Comment by amdvlk-admin** [comment]
### Test summary for commit fba4b5319533cdba5b1dd00b288c023248d3172f
<details><summary>CTS tests (Failed: 0/282025)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 38862/70507 (55.1%)</li>
<li>Failed: 0/70507 (0.0%)</li>
<li>Not Supported: 31645/70507 (44.9%)</li><li>Warnings: 0/70507 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi2x</summary><ul>

<li>Passed: 39112/70507 (55.5%)</li>
<li>Failed: 0/70507 (0.0%)</li>
<li>Not Supported: 31395/70507 (44.5%)</li><li>Warnings: 0/70507 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35959/70509 (51.0%)</li>
<li>Failed: 0/70509 (0.0%)</li>
<li>Not Supported: 34550/70509 (49.0%)</li><li>Warnings: 0/70509 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 35861/70502 (50.9%)</li>
<li>Failed: 0/70502 (0.0%)</li>
<li>Not Supported: 34641/70502 (49.1%)</li><li>Warnings: 0/70502 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**